Add conv1d to burn-import (onnx)

mrvollger opened the pull request:
## Pull Request Template

### Checklist

- [ x ] Confirm that `run-checks.sh` has been executed.
I have run this and nearly all tests pass except for ones associated with `burn_wgpu`. However I dont think this has anything to do with the new code and just my system. To confirm this I checked and the same tests fail if I switch to the main branch without my updates, so I think it is okay.

### Changes

I added a Conv1d as a new operation to burn-import. I largely followed  Conv2d as an example and followed the instructions in the README and testing README. However, I am a first time contributor so any extra review would be much appreciated! 

### Testing

I have added a test case to onnx-tests and confirmed that the results from pytorch match the results of burn-import. To ensure the results matched I had to set the random seed for pytorch. I noticed this was missing from the Conv2d python test script which caused to onnx test to fail so I added the forced seed and updated the expected sum in Conv2d. 


Thanks in advance! 
Mitchell

antimora commented:
Thank you, @mrvollger, for adding this implementation. I will definitely review it and get back to you. 

So you know, I have been fixing issues with burn-import. I have submitted a couple PRs (pending) and one unsubmitted yet (regarding Reshape). Depending how it goes you may have to merge these changes or we can your changes first, and I'll have to remerge mine. 

I am happy to see more contributors. I have been fixing some bugs when I was importing squeezenet.onnx. Happy to report, I am generating Burn model. As soon as my PRs are merged, I'll submit the model to https://github.com/burn-rs/models.

antimora commented on burn-import/src/onnx/op_configuration.rs:
```diff
@@ -28,6 +28,48 @@ pub fn attr_value_f32(value: &AttributeValue, target: &mut f32) {
 }
 
+/// Create a Conv1dConfig from the attributes of the node
+/// TODO check this function carefully, I don't understand it well enough
```
We need to remove this TODO before merging.

antimora commented on burn-import/src/onnx/op_configuration.rs:
```diff
@@ -28,6 +28,48 @@ pub fn attr_value_f32(value: &AttributeValue, target: &mut f32) {
+    let shape = tensor.shape.unwrap();
+    let channels_in = shape[1];
+    let channels_out = shape[0];
+    //let channels: [usize; 2] = [shape[1], shape[0]];
```
Dead code should be removed.

antimora commented on burn-import/src/onnx/op_configuration.rs:
```diff
@@ -357,3 +399,41 @@ fn padding_config(pads: &[i64]) -> PaddingConfig2d {
+        // i.e [0, 0]
+        PaddingConfig1d::Valid
+    } else if left == right {
+        // i.e [2, 3, 2, 3]
```
` // i.e [2, 3, 2, 3]` should be ` // i.e [2, 2] to be accurate.

antimora commented on burn-import/src/onnx/op_configuration.rs:
```diff
@@ -357,3 +399,41 @@ fn padding_config(pads: &[i64]) -> PaddingConfig2d {
+        PaddingConfig1d::Valid
+    } else if left == right {
+        // i.e [2, 3, 2, 3]
+        PaddingConfig1d::Explicit(left as usize) //#(left as usize, top as usize)
```
dead code `//#(left as usize, top as usize)` should be removed

antimora commented on burn-import/src/onnx/to_burn.rs:
```diff
@@ -453,6 +455,23 @@ impl ONNXGraph {
         )
     }
 
+    /// TODO check that this is valid, I am not 100% sure on it.
```
This todo comment should be removed.

antimora commented on burn-import/src/onnx/dim_inference.rs:
```diff
@@ -297,6 +298,23 @@ fn flatten_update_outputs(node: &mut Node) {
 }
 
+/// Infers the shape of a Conv1d node and replaces the shape of the output tensor.
+///
+/// The shape of the output tensor is calculated by running the actual convolution operation.
```
The comment is outdated and applicable anymore. If you see it on Conv2d, please remove it as well.

Before a big refactor, we used to calculate the output shape by calling the Conv2d code. But we have figured out we do not need to know the shapes. But instead only a dimension of the output is needed. So this comment was forgotten to be cleaned (and probably confusing others now =D)

mrvollger replied:
Done, I removed it on both functions!

antimora reviewed (changes requested):
You have everything got it correctly. And I am confident because your tests. The onnx file conversation and running guarantees the code's syntax correctness and the actual execution is correct. Even weights and biases are loaded correctly. 

Thanks for adding the seed. It's correct to add it since the weights and biases are initialized from a random data.

antimora commented:
If we could get the comments fixed, we can merge the PR before mine. I'll take care of the conflicts.

mrvollger commented:
Awesome, I will get on these requests!

mrvollger commented:
I am very confused why the tests now fail... They run locally and I only made comment/spacing changes in addressing the review. I will investigate a bit more.

antimora commented:
> I am very confused why the tests now fail... They run locally and I only made comment/spacing changes in addressing the review. I will investigate a bit more.

The CI is failing because of something else. The clippy got upgraded. We will fix and rerun the tests.

antimora reviewed (approved):
Looks good! Thank for you contribution!

Ready to merge.
